Upgrades for mdbook 0.2.0, and link style cleanup.

alercah opened the pull request:
See individual commit messages for description of what I did.

I have not verified all links. `mdbook-linkcheck` does not yet support 0.2.0.

Havvy commented on src/items/traits.md:
```diff
@@ -3,9 +3,9 @@
-- [functions](items/associated-items.html#associated-functions-and-methods)
-- [types](items/associated-items.html#associated-types)
-- [constants](items/associated-items.html#associated-constants)
+- [functions]
```
I feel like this is going to cause confusion as just being functions. How about `[function][associated functions]`?

alercah replied:
I don't want to make functional changes here.

Havvy commented on src/items/type-aliases.md:
```diff
@@ -29,7 +27,8 @@ let _: F = E::A;  // OK
+[type]: ../types.html
+[IDENTIFIER]: ../identifiers.html
+[_Generics_]: generics.html
+[_WhereClause_]: html#where-clauses
```
Missing the `generics.`.

Havvy commented on src/items/unions.md:
```diff
@@ -143,9 +143,10 @@ generics, trait implementations, inherent implementations, coherence, pattern
-[_StructFields_]: items/structs.html
+[RFC 1897 "Unions v1.2"]: https://github.com/rust-lang/rfcs/pull/1897
+[IDENTIFIER]: ../identifiers.html
+[_Generics_]: html
```
Missing `generics.` and similar for next two lines.

Havvy commented on src/expressions.md:
```diff
@@ -233,9 +233,9 @@ Implicit borrows may be taken in the following expressions:
 Certain types of expressions can be evaluated at compile time. These are called
 _constant expressions_. Certain places, such as in
-[constants](items/constant-items.html) and [statics](items/static-items.html),
+[constants] and [statics](static variables),
```
Square brackets, not round brackets.

Havvy commented on src/items.md:
```diff
@@ -36,3 +33,19 @@ as if the item was declared outside the scope &mdash; it is still a static item
+
+[modules]: items/modules.html
+[paths]: paths.html
+[modules]: items/modules.html
```
modules is duplicated.

Havvy commented on src/types.md:
```diff
@@ -742,3 +720,30 @@ impl Printable for String {
+[structs]: #struct-types
+[enumerations]: #enumerated-types
+[unions]: #union-types
+[type definitions]: ../grammar.html#type-definitions
```
Do we want to be linking to that grammar still?

Havvy commented:
Oh, I commented on commits as I checked on each commit. Thus some stuff says out of date even though it's not.

Havvy commented:
This needs to be rebased pretty badly. I think it'd be easier if the alphabetizing commit was its own PR.

alercah commented:
Rebased and updated, removed the sorting/organization and just stuck on the version upgrade.

alercah commented:
@steveklabnik

ehuss commented:
These links don't look correct to me.  They should all be relative to the page they are on.  For example, `introduction.md` has a bunch of `../` links that should not have `../`.  If you visit the page [here on Github](https://github.com/rust-lang-nursery/reference/blob/198488b45afdd16cd0dd8194bb7d3703d19a5ec1/src/introduction.md), none of the links are working.

There also looks to be some unintended changes to `//` comments in `operator-expr.md`.

ehuss commented:
Oh, `.travis.yml` will need to be updated, too.

steveklabnik commented:
Oh shucks, it looks like the upstream PR for multi-mdbook support has not actually landed. I thought it had. Let me take care of that ASAP.

steveklabnik commented:
Now r+'d https://github.com/rust-lang/rust/pull/57495

alercah commented:
@ehuss Ah right, I thought I reverted all the ones outside of subdirs since I did them by mistake, but missed `introduction.md`. I fixed some I missed too that linkcheck found.

ehuss commented on src/expressions/array-expr.md:
```diff
@@ -81,11 +80,12 @@ arr[10];                  // warning: index out of bounds
 by implementing the [Index] and [IndexMut] traits.
 
+[array]: ../types/array.html
 [IndexMut]: ../std/ops/trait.IndexMut.html
```
These two std links need to be fixed up.

ehuss commented on .travis.yml:
```diff
@@ -4,7 +4,7 @@ rust:
 
 install:
-  - travis_retry curl -Lf https://github.com/rust-lang-nursery/mdBook/releases/download/v0.1.7/mdbook-v0.1.7-x86_64-unknown-linux-gnu.tar.gz | tar -xz --directory=$HOME/.cargo/bin
+  - travis_retry curl -Lf https://github.com/rust-lang-nursery/mdBook/releases/download/v0.2.1/mdbook-v0.2.1-x86_64-unknown-linux-gnu.tar.gz | tar -xz --directory=$HOME/.cargo/bin
```
Oh, hmm, this needs to be 0.2.3.  @steveklabnik would it be possible to publish a binary for the latest release on github?

steveklabnik commented:
So, I have good news, and I have bad news. Good news is, I ran this against upstream's linkchecker. Bad news is that I got:

<details>
```text
reference\expressions\array-expr.html:225: broken link - reference\std\ops\trait.Index.html
reference\expressions\array-expr.html:225: broken link - reference\std\ops\trait.IndexMut.html
reference\print.html:2243: broken link - identifiers.html
reference\print.html:2244: broken link - identifiers.html
reference\print.html:2245: broken link - attributes.html
reference\print.html:2246: broken link - items.html
reference\print.html:2249: broken link - items.html
reference\print.html:2280: broken link - paths.html
reference\print.html:2345: broken link - crates-and-source-files.html
reference\print.html:2347: broken link - crates-and-source-files.html
reference\print.html:2348: broken link - attributes.html
reference\print.html:2353: broken link - conditional-compilation.html
reference\print.html:2354: broken link - attributes.html
reference\print.html:2354: broken link - attributes.html
reference\print.html:2354: broken link - attributes.html
reference\print.html:2362: broken link - identifiers.html
reference\print.html:2364: broken link - identifiers.html
reference\print.html:2367: broken link - identifiers.html
reference\print.html:2376: broken link - identifiers.html
reference\print.html:2400: broken link - crates-and-source-files.html
reference\print.html:2401: broken link - items\modules.html
reference\print.html:2405: broken link - items\use-declarations.html
reference\print.html:2407: broken link - items\use-declarations.html
reference\print.html:2432: broken link - attributes.html
reference\print.html:2440: broken link - paths.html
reference\print.html:2441: broken link - paths.html
reference\print.html:2442: broken link - paths.html
reference\print.html:2442: broken link - identifiers.html

reference\print.html:6145: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.Fn.html
reference\print.html:6145: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnMut.html
reference\print.html:6146: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnOnce.html

reference\print.html:8032: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.Fn.html
reference\print.html:8032: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnMut.html
reference\print.html:8032: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnOnce.html
reference\print.html:8032: broken link - special-types-and-traits.html
reference\print.html:8033: broken link - special-types-and-traits.html
reference\print.html:8033: broken link - special-types-and-traits.html
reference\print.html:8033: broken link - special-types-and-traits.html
reference\print.html:8035: broken link - expressions\closure-expr.html
reference\print.html:8134: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnOnce.html
reference\print.html:8140: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnMut.html
reference\print.html:8144: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.Fn.html
reference\print.html:8148: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.Fn.html
reference\print.html:8148: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnMut.html

reference\print.html:7914: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\vec\struct.Vec.html

```
</details>

... and more, but I won't paste the same line about print.html over and over again...

I am very confused about `print.html`, because I did https://github.com/rust-lang-nursery/mdBook/pull/866 to fix this.

steveklabnik commented:
> All of the links to .html in-repo need to be updated to .md. This allows links to work on GitHub, and mdbook doesn't handle the .html links properly in the print page (all links not in the root are currently broken on the print page).

ahhhhhh yes

steveklabnik commented:
The upstream PR has landed! @alercah let me know when you want me to try the linkchecker again :)

alercah commented:
Apologies for the delay here, had some issues. Should be good to go now, though I can't get linkcheck to ignore the parent directories so I can't be totally sure.

ehuss commented:
Hm, something's wrong. I swore I tested the print page on 0.2.3, but there are still problems.  There are two issues:
- Links like `items/unions.md` jump out of the print page.  I would assume the 0.1 behavior should be retained where the print page links to itself.
- Links like `../../std/mem/fn.discriminant.html` are not rebased when used from a subdirectory, and are broken.

I'll see if I can figure out what's wrong in the next few days.

ehuss commented:
I have posted a proposed fix for the print.html links at https://github.com/rust-lang-nursery/mdBook/pull/871.

As for publishing 0.2.3, it looks like the deploy scripts are broken ([travis](https://travis-ci.com/rust-lang-nursery/mdBook/builds/97850452) and [appveyor](https://ci.appveyor.com/project/rust-lang-libs/mdbook/builds/21723156/job/w0ebv0gtfwes4voq)).  I have posted a partial fix for Travis at https://github.com/rust-lang-nursery/mdBook/pull/872.  The API keys need to be fixed since the transition to travis.com.  @steveklabnik if you have permissions, can you take care of that? Or if not, I can try to bug infra and get things worked out.  Alternatively the releases can be made manually (it's just a compressed copy of the mdbook executable), and uploaded manually on the GitHub releases page.

mati865 commented:
There is mdbook 0.3 so this should be unblocked but there are so many conflicts and new links this PR probably has to be redone from scratch.

ehuss commented:
@mati865 There was another bug blocking the update of the reference and cargo (https://github.com/rust-lang-nursery/mdBook/pull/963).  It is now fixed, I just need to make a new release.  I'll be posting a new PR here to replace this one when things are ready.  It will require a coordinated update due to some changes to anchor tags.  I think I'll be able to eradicate 0.1 during that update.

ehuss commented:
I'm going to close this for now.  The reference has been updated to mdbook 0.3 in #641.  If there are any link cleanups we want to do, we can always do them in future PRs. Thanks @alercah, sorry it took so long to iron out all the kinks in mdbook to make this work.
